File: src/accessiweather/current_location.py
"""One-time native current-location detection helpers."""

from __future__ import annotations

import asyncio
import importlib
import logging
import sys
import threading
from dataclasses import dataclass
from enum import Enum
from typing import Any, Protocol

from .location_classification import is_us_location
from .models import Location

logger = logging.getLogger(__name__)


class LocationDetectionStatus(Enum):
    """Outcome of a one-time current-location request."""

    SUCCESS = "success"
    DENIED = "denied"
    TIMEOUT = "timeout"
    UNAVAILABLE = "unavailable"
    UNSUPPORTED = "unsupported"


@dataclass(frozen=True)
class CurrentCoordinates:
    """Coordinates returned by a native location provider."""

    latitude: float
    longitude: float
    accuracy_meters: float | None = None


@dataclass(frozen=True)
class CurrentLocationResult:
    """Normalized result for the UI."""

    status: LocationDetectionStatus
    message: str
    coordinates: CurrentCoordinates | None = None
    location: Location | None = None


class CurrentLocationError(Exception):
    """Expected provider failure normalized for UI handling."""

    def __init__(self, status: LocationDetectionStatus, message: str) -> None:
        """Store the normalized status alongside the user-facing message."""
        super().__init__(message)
        self.status = status
        self.message = message


class CurrentLocationProvider(Protocol):
    """Protocol implemented by native one-shot providers."""

    async def detect(self, timeout_seconds: float = 15.0) -> CurrentCoordinates:
        """Return current coordinates or raise ``CurrentLocationError``."""
        raise NotImplementedError


def location_from_coordinates(
    coordinates: CurrentCoordinates,
    *,
    name: str | None = None,
) -> Location:
    """Create the same editable Location object used by manual location saves."""
    label = name or f"Current Location ({coordinates.latitude:.4f}, {coordinates.longitude:.4f})"
    location = Location(name=label, latitude=coordinates.latitude, longitude=coordinates.longitude)
    if is_us_location(location):
        location.country_code = "US"
    return location


class CurrentLocationService:
    """Facade that normalizes native provider outcomes for the UI."""

    def __init__(self, provider: CurrentLocationProvider | None = None) -> None:
        """Initialize with an injected provider or the native provider for this platform."""
        self._provider = provider or get_native_provider()

    async def detect_once(self, timeout_seconds: float = 15.0) -> CurrentLocationResult:
        """Run one user-initiated location request."""
        try:
            coordinates = await self._provider.detect(timeout_seconds=timeout_seconds)
        except CurrentLocationError as exc:
            return CurrentLocationResult(status=exc.status, message=exc.message)
        except TimeoutError:
            return CurrentLocationResult(
                status=LocationDetectionStatus.TIMEOUT,
                message="Current location detection timed out. You can still search manually.",
            )
        except Exception as exc:  # noqa: BLE001 - native providers can fail many ways
            logger.debug("Current location detection failed unexpectedly: %s", exc, exc_info=True)
            return CurrentLocationResult(
                status=LocationDetectionStatus.UNAVAILABLE,
                message="Current location is unavailable. You can still search manually.",
            )

        location = location_from_coordinates(coordinates)
        return CurrentLocationResult(
            status=LocationDetectionStatus.SUCCESS,
            message="Current location detected. Review the editable name before saving.",
            coordinates=coordinates,
            location=location,
        )


class UnsupportedLocationProvider:
    """Provider used where native location detection is not implemented."""

    async def detect(self, timeout_seconds: float = 15.0) -> CurrentCoordinates:
        raise CurrentLocationError(
            LocationDetectionStatus.UNSUPPORTED,
            "Current location detection is not supported on this platform. You can still search manually.",
        )


class WindowsLocationProvider:
    """One-shot Windows Location Services provider through WinRT when available."""

    async def detect(self, timeout_seconds: float = 15.0) -> CurrentCoordinates:
        try:
            geolocation = importlib.import_module("winsdk.windows.devices.geolocation")
        except Exception as exc:  # noqa: BLE001
            raise CurrentLocationError(
                LocationDetectionStatus.UNAVAILABLE,
                "Windows Location Services support is not installed. You can still search manually.",
            ) from exc

        try:
            access = await asyncio.wait_for(
                geolocation.Geolocator.request_access_async(),
                timeout=timeout_seconds,
            )
        except TimeoutError as exc:
            raise CurrentLocationError(
                LocationDetectionStatus.TIMEOUT,
                "Windows Location Services did not respond. You can still search manually.",
            ) from exc

        access_name = str(getattr(access, "name", access)).lower()
        if "allowed" not in access_name:
            raise CurrentLocationError(
                LocationDetectionStatus.DENIED,
                "Location permission was denied. You can still search manually.",
            )

        locator = geolocation.Geolocator()
        if hasattr(geolocation, "PositionAccuracy"):
            locator.desired_accuracy = geolocation.PositionAccuracy.DEFAULT

        try:
            position = await asyncio.wait_for(
                locator.get_geoposition_async(),
                timeout=timeout_seconds,
            )
        except TimeoutError as exc:
            raise CurrentLocationError(
                LocationDetectionStatus.TIMEOUT,
                "Windows Location Services timed out. You can still search manually.",
            ) from exc

        return _coordinates_from_windows_position(position)


def _coordinates_from_windows_position(position: Any) -> CurrentCoordinates:
    """Extract latitude/longitude from common WinRT Python projections."""
    coordinate = getattr(position, "coordinate", None)
    point = getattr(coordinate, "point", None)
    point_position = getattr(point, "position", None)

    latitude = getattr(point_position, "latitude", None)
    longitude = getattr(point_position, "longitude", None)
    if latitude is None or longitude is None:
        latitude = getattr(coordinate, "latitude", None)
        longitude = getattr(coordinate, "longitude", None)

    if latitude is None or longitude is None:
        raise CurrentLocationError(
            LocationDetectionStatus.UNAVAILABLE,
            "Windows Location Services returned no coordinates. You can still search manually.",
        )

    accuracy = getattr(coordinate, "accuracy", None)
    return CurrentCoordinates(
        latitude=float(latitude),
        longitude=float(longitude),
        accuracy_meters=float(accuracy) if accuracy is not None else None,
    )


class MacOSLocationProvider:
    """One-shot CoreLocation provider through PyObjC when available."""

    async def detect(self, timeout_seconds: float = 15.0) -> CurrentCoordinates:
        try:
            CoreLocation = importlib.import_module("CoreLocation")
            Foundation = importlib.import_module("Foundation")
        except Exception as exc:  # noqa: BLE001
            raise CurrentLocationError(
                LocationDetectionStatus.UNAVAILABLE,
                "macOS CoreLocation support is not installed. You can still search manually.",
            ) from exc

        loop = asyncio.get_running_loop()
        future: asyncio.Future[CurrentCoordinates] = loop.create_future()

        class _LocationDelegate(Foundation.NSObject):
            def locationManager_didUpdateLocations_(self, manager, locations):  # noqa: N802
                latest = locations.lastObject()
                coordinate = latest.coordinate()
                accuracy = latest.horizontalAccuracy()
                if not future.done():
                    loop.call_soon_threadsafe(
                        future.set_result,
                        CurrentCoordinates(
                            latitude=float(coordinate.latitude),
                            longitude=float(coordinate.longitude),
                            accuracy_meters=float(accuracy) if accuracy >= 0 else None,
                        ),
                    )
                manager.stopUpdatingLocation()

            def locationManager_didFailWithError_(self, manager, error):  # noqa: N802
                if not future.done():
                    loop.call_soon_threadsafe(
                        future.set_exception,
                        CurrentLocationError(
                            LocationDetectionStatus.UNAVAILABLE,
                            "macOS could not detect your current location. You can still search manually.",
                        ),
                    )
                manager.stopUpdatingLocation()

            def locationManagerDidChangeAuthorization_(self, manager):  # noqa: N802
                status = manager.authorizationStatus()
                denied_statuses = {
                    getattr(CoreLocation, "kCLAuthorizationStatusDenied", object()),
                    getattr(CoreLocation, "kCLAuthorizationStatusRestricted", object()),
                }
                if status in denied_statuses and not future.done():
                    loop.call_soon_threadsafe(
                        future.set_exception,
                        CurrentLocationError(
                            LocationDetectionStatus.DENIED,
                            "Location permission was denied. You can still search manually.",
                        ),
                    )

        manager = CoreLocation.CLLocationManager.alloc().init()
        delegate = _LocationDelegate.alloc().init()
        manager.setDelegate_(delegate)

        def _run_core_location() -> None:
            manager.requestWhenInUseAuthorization()
            if hasattr(manager, "requestLocation"):
                manager.requestLocation()
            else:
                manager.startUpdatingLocation()
            Foundation.NSRunLoop.currentRunLoop().runUntilDate_(
                Foundation.NSDate.dateWithTimeIntervalSinceNow_(timeout_seconds)
            )

        thread = threading.Thread(target=_run_core_location, daemon=True)
        thread.start()

        try:
            return await asyncio.wait_for(future, timeout=timeout_seconds + 0.5)
        except TimeoutError as exc:
            manager.stopUpdatingLocation()
            raise CurrentLocationError(
                LocationDetectionStatus.TIMEOUT,
                "macOS location detection timed out. You can still search manually.",
            ) from exc
        finally:
            # Keep delegate alive until the future completes; then break the reference cycle.
            manager.setDelegate_(None)


def get_native_provider() -> CurrentLocationProvider:
    """Return the provider for the current platform."""
    if sys.platform == "win32":
        return WindowsLocationProvider()
    if sys.platform == "darwin":
        return MacOSLocationProvider()
    return UnsupportedLocationProvider()

File: rust/crates/aw-providers/src/current_location.rs
//! One-time "Use my current location" detection, ported from
//! `accessiweather.current_location`. Windows uses Windows Location Services
//! (WinRT `Geolocator`); other platforms report "unsupported" (see the port
//! notes: the macOS CoreLocation provider is not implemented yet).

use std::time::Duration;

use aw_core::location::is_us_location;
use aw_core::model::Location;

/// Outcome of a one-time current-location request.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum LocationDetectionStatus {
    Success,
    Denied,
    Timeout,
    Unavailable,
    Unsupported,
}

impl LocationDetectionStatus {
    pub fn value(self) -> &'static str {
        match self {
            Self::Success => "success",
            Self::Denied => "denied",
            Self::Timeout => "timeout",
            Self::Unavailable => "unavailable",
            Self::Unsupported => "unsupported",
        }
    }
}

/// Coordinates returned by a native location provider.
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct CurrentCoordinates {
    pub latitude: f64,
    pub longitude: f64,
    pub accuracy_meters: Option<f64>,
}

/// Normalized result for the UI.
#[derive(Debug, Clone, PartialEq)]
pub struct CurrentLocationResult {
    pub status: LocationDetectionStatus,
    pub message: String,
    pub coordinates: Option<CurrentCoordinates>,
    pub location: Option<Location>,
}

/// Provider failure: an expected outcome with its user-facing message, a
/// bare timeout, or anything else.
#[derive(Debug, Clone, PartialEq)]
pub enum DetectError {
    Known(LocationDetectionStatus, String),
    Timeout,
    Failed(String),
}

pub trait CurrentLocationProvider: Send + Sync {
    fn detect(&self, timeout: Duration) -> Result<CurrentCoordinates, DetectError>;
}

pub const DEFAULT_TIMEOUT: Duration = Duration::from_secs(15);

/// The editable location a manual save would create for the coordinates.
pub fn location_from_coordinates(coordinates: CurrentCoordinates, name: Option<&str>) -> Location {
    let label = match name.filter(|n| !n.is_empty()) {
        Some(n) => n.to_string(),
        None => format!(
            "Current Location ({:.4}, {:.4})",
            coordinates.latitude, coordinates.longitude
        ),
    };
    let mut location = Location::new(label, coordinates.latitude, coordinates.longitude);
    if is_us_location(&location) {
        location.country_code = Some("US".into());
    }
    location
}

/// Facade that normalizes native provider outcomes for the UI.
pub struct CurrentLocationService {
    provider: Box<dyn CurrentLocationProvider>,
}

impl Default for CurrentLocationService {
    fn default() -> Self {
        Self::new(native_provider())
    }
}

impl CurrentLocationService {
    pub fn new(provider: Box<dyn CurrentLocationProvider>) -> Self {
        Self { provider }
    }

    /// Run one user-initiated location request (blocking; call off the UI thread).
    pub fn detect_once(&self, timeout: Duration) -> CurrentLocationResult {
        let failure = |status, message: &str| CurrentLocationResult {
            status,
            message: message.to_string(),
            coordinates: None,
            location: None,
        };
        match self.provider.detect(timeout) {
            Ok(coordinates) => CurrentLocationResult {
                status: LocationDetectionStatus::Success,
                message: "Current location detected. Review the editable name before saving.".into(),
                coordinates: Some(coordinates),
                location: Some(location_from_coordinates(coordinates, None)),
            },
            Err(DetectError::Known(status, message)) => failure(status, &message),
            Err(DetectError::Timeout) => failure(
                LocationDetectionStatus::Timeout,
                "Current location detection timed out. You can still search manually.",
            ),
            Err(DetectError::Failed(reason)) => {
                tracing::debug!("Current location detection failed unexpectedly: {reason}");
                failure(
                    LocationDetectionStatus::Unavailable,
                    "Current location is unavailable. You can still search manually.",
                )
            }
        }
    }
}

/// Provider used where native location detection is not implemented.
pub struct UnsupportedLocationProvider;

impl CurrentLocationProvider for UnsupportedLocationProvider {
    fn detect(&self, _timeout: Duration) -> Result<CurrentCoordinates, DetectError> {
        Err(DetectError::Known(
            LocationDetectionStatus::Unsupported,
            "Current location detection is not supported on this platform. You can still search manually."
                .into(),
        ))
    }
}

/// The provider for this platform.
pub fn native_provider() -> Box<dyn CurrentLocationProvider> {
    #[cfg(windows)]
    {
        Box::new(windows_provider::WindowsLocationProvider)
    }
    #[cfg(not(windows))]
    {
        Box::new(UnsupportedLocationProvider)
    }
}

#[cfg(windows)]
pub use windows_provider::WindowsLocationProvider;

#[cfg(windows)]
mod windows_provider {
    use std::sync::mpsc;
    use std::time::Duration;

    use windows::Devices::Geolocation::{GeolocationAccessStatus, Geolocator, PositionAccuracy};
    use windows_future::IAsyncOperation;

    use super::{CurrentCoordinates, CurrentLocationProvider, DetectError, LocationDetectionStatus};

    /// One-shot Windows Location Services provider.
    pub struct WindowsLocationProvider;

    fn failed(e: windows::core::Error) -> DetectError {
        DetectError::Failed(e.to_string())
    }

    /// Wait for a WinRT operation, cancelling it on timeout (`None`).
    fn wait<T>(op: IAsyncOperation<T>, timeout: Duration) -> Result<Option<T>, DetectError>
    where
        T: windows::core::RuntimeType + Send + 'static,
    {
        let (tx, rx) = mpsc::channel();
        op.when(move |result| {
            let _ = tx.send(result);
        })
        .map_err(failed)?;
        match rx.recv_timeout(timeout) {
            Ok(result) => result.map(Some).map_err(failed),
            Err(_) => {
                let _ = op.Cancel();
                Ok(None)
            }
        }
    }

    impl CurrentLocationProvider for WindowsLocationProvider {
        fn detect(&self, timeout: Duration) -> Result<CurrentCoordinates, DetectError> {
            let access = wait(Geolocator::RequestAccessAsync().map_err(failed)?, timeout)?
                .ok_or_else(|| {
                    DetectError::Known(
                        LocationDetectionStatus::Timeout,
                        "Windows Location Services did not respond. You can still search manually."
                            .into(),
                    )
                })?;
            if access != GeolocationAccessStatus::Allowed {
                return Err(DetectError::Known(
                    LocationDetectionStatus::Denied,
                    "Location permission was denied. You can still search manually.".into(),
                ));
            }

            let locator = Geolocator::new().map_err(failed)?;
            locator
                .SetDesiredAccuracy(PositionAccuracy::Default)
                .map_err(failed)?;
            let position = wait(locator.GetGeopositionAsync().map_err(failed)?, timeout)?
                .ok_or_else(|| {
                    DetectError::Known(
                        LocationDetectionStatus::Timeout,
                        "Windows Location Services timed out. You can still search manually."
                            .into(),
                    )
                })?;

            let no_coordinates = || {
                DetectError::Known(
                    LocationDetectionStatus::Unavailable,
                    "Windows Location Services returned no coordinates. You can still search manually."
                        .into(),
                )
            };
            let coordinate = position.Coordinate().map_err(|_| no_coordinates())?;
            let point = coordinate
                .Point()
                .and_then(|p| p.Position())
                .map_err(|_| no_coordinates())?;
            Ok(CurrentCoordinates {
                latitude: point.Latitude,
                longitude: point.Longitude,
                accuracy_meters: coordinate.Accuracy().ok(),
            })
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    struct Fixed(Result<CurrentCoordinates, DetectError>);

    impl CurrentLocationProvider for Fixed {
        fn detect(&self, _timeout: Duration) -> Result<CurrentCoordinates, DetectError> {
            self.0.clone()
        }
    }

    fn service(result: Result<CurrentCoordinates, DetectError>) -> CurrentLocationService {
        CurrentLocationService::new(Box::new(Fixed(result)))
    }

    #[test]
    fn unsupported_provider_reports_manual_fallback() {
        let r = CurrentLocationService::new(Box::new(UnsupportedLocationProvider)).detect_once(DEFAULT_TIMEOUT);
        assert_eq!(r.status, LocationDetectionStatus::Unsupported);
        assert_eq!(
            r.message,
            "Current location detection is not supported on this platform. You can still search manually."
        );
    }

    #[test]
    fn failures_are_normalized() {
        let r = service(Err(DetectError::Timeout)).detect_once(DEFAULT_TIMEOUT);
        assert_eq!(r.status, LocationDetectionStatus::Timeout);
        assert_eq!(r.message, "Current location detection timed out. You can still search manually.");
        let r = service(Err(DetectError::Failed("boom".into()))).detect_once(DEFAULT_TIMEOUT);
        assert_eq!(r.status, LocationDetectionStatus::Unavailable);
        assert_eq!(r.message, "Current location is unavailable. You can still search manually.");
        let denied = DetectError::Known(LocationDetectionStatus::Denied, "no".into());
        assert_eq!(service(Err(denied)).detect_once(DEFAULT_TIMEOUT).message, "no");
    }

    #[test]
    fn coordinates_become_an_editable_location() {
        let coords = CurrentCoordinates { latitude: 40.7128, longitude: -74.006, accuracy_meters: Some(25.0) };
        let r = service(Ok(coords)).detect_once(DEFAULT_TIMEOUT);
        assert_eq!(r.status, LocationDetectionStatus::Success);
        assert_eq!(r.message, "Current location detected. Review the editable name before saving.");
        let location = r.location.unwrap();
        assert_eq!(location.name, "Current Location (40.7128, -74.0060)");
        assert_eq!(location.country_code.as_deref(), Some("US"));

        let toronto = CurrentCoordinates { latitude: 43.65, longitude: -79.38, accuracy_meters: None };
        assert_eq!(location_from_coordinates(toronto, None).country_code, None);
    }
}
